File: dewicom-desktop/leader-election.js
/**
 * Élection de leader DewiCom — côté Node.js/Electron.
 * Même protocole que LeaderElection.java (Bully sur multicast UDP port 9998).
 *
 * Messages : ELECTION:<nodeId>:<ip>  LEADER:<nodeId>:<ip>  HEARTBEAT:<nodeId>:<ip>
 */
const dgram = require("dgram");
const os    = require("os");
const fs    = require("fs");
const path  = require("path");

const MCAST_ADDR         = "224.0.0.251";
const ELECT_PORT         = 9998;
const HEARTBEAT_INTERVAL = 1000;  // heartbeat toutes les 1s
const LEADER_TIMEOUT     = 3000;  // 3s sans heartbeat → re-élection
const ELECTION_WAIT      = 1000;  // 1s d'attente avant de se proclamer leader
const BROADCAST_COOLDOWN = 300;   // anti-storm : cooldown entre deux broadcasts ELECTION

function getForcedInterface() {
  try {
    const { app } = require("electron");
    const configPath = path.join(app.getPath("userData"), "server-config.json");
    const saved = JSON.parse(fs.readFileSync(configPath, "utf8"));
    return saved.forcedInterface || null;
  } catch (e) { return null; }
}

function getLocalIP() {
  const forced = getForcedInterface();
  if (forced) return forced;
  const candidates = [];
  for (const [name, ifaces] of Object.entries(os.networkInterfaces())) {
    for (const iface of ifaces) {
      if (iface.family !== "IPv4" || iface.internal) continue;
      if (iface.address.startsWith("169.254.")) continue; // APIPA — pas de routeur
      const lname = name.toLowerCase();
      let score = 0;
      if (lname.includes("virtualbox") || lname.includes("vmware") ||
          lname.includes("vbox") || lname.includes("hyper-v") ||
          lname.includes("loopback") || lname.includes("tap") ||
          lname.includes("tun") || lname.includes("docker") ||
          lname.startsWith("virbr") || lname.startsWith("veth") ||
          lname.startsWith("br-") || lname.startsWith("lxc") ||
          lname.startsWith("lxd")) score -= 10;
      if (iface.address.startsWith("192.168.") || iface.address.startsWith("10.") ||
          iface.address.startsWith("172.")) score += 5;
      candidates.push({ address: iface.address, score });
    }
  }
  if (candidates.length === 0) return "127.0.0.1";
  candidates.sort((a, b) => b.score - a.score);
  return candidates[0].address;
}

function ipToNodeId(ip) {
  const p = ip.split(".").map(Number);
  const base = ((p[0] & 0xFF) * 16777216) + ((p[1] & 0xFF) * 65536) + ((p[2] & 0xFF) * 256) + (p[3] & 0xFF);
  // Le desktop a toujours une priorité supérieure aux APK (offset de 2^32)
  return base + 4294967296;
}

class LeaderElection {
  constructor(callbacks = {}) {
    this.myIP      = getLocalIP();
    this.myNodeId  = ipToNodeId(this.myIP);
    this.state     = "FOLLOWER"; // FOLLOWER | CANDIDATE | LEADER
    this.leaderIP  = null;
    this.lastHeartbeat = 0;
    this.running   = false;

    this.onBecomeLeader  = callbacks.onBecomeLeader  || (() => {});
    this.onLeaderElected = callbacks.onLeaderElected || (() => {});

    this._heartbeatTimer  = null;
    this._watchdogTimer   = null;
    this._electionTimer   = null;
    this._electionPending = false; // debounce : une seule élection à la fois
    this._lastBroadcastTs = 0;     // anti-storm : timestamp du dernier broadcast ELECTION

    console.log(`[election] Init — IP: ${this.myIP}, nodeId: ${this.myNodeId}`);
  }

  start() {
    this.running = true;
    this._openSocket();
    const delay = 500 + Math.random() * 1000;
    setTimeout(() => this._startElection(), delay);
  }

  stop() {
    this.running = false;
    clearInterval(this._heartbeatTimer);
    clearInterval(this._watchdogTimer);
    clearTimeout(this._electionTimer);
    this._electionPending = false;
    try { this._socket?.close(); } catch (e) {}
  }

  // Pause propre avant veille : arrête les timers sans fermer le socket
  pauseTimers() {
    clearInterval(this._heartbeatTimer);  this._heartbeatTimer = null;
    clearInterval(this._watchdogTimer);   this._watchdogTimer  = null;
    clearTimeout(this._electionTimer);    this._electionTimer  = null;
    this._electionPending = false;
    console.log("[election] Timers suspendus (veille système)");
  }

  // Reprise après réveil : repart proprement selon l'état courant
  resumeTimers() {
    if (!this.running) return;
    this.lastHeartbeat = Date.now(); // évite un faux timeout immédiat
    if (this.state === "LEADER") {
      this._startHeartbeat(); // reprend uniquement le heartbeat — pas de callback onBecomeLeader
    } else {
      this._startWatchdog(); // reprend watchdog follower
    }
    console.log("[election] Timers repris après réveil (état: " + this.state + ")");
  }

  isLeader()   { return this.state === "LEADER"; }
  getLeaderIP(){ return this.leaderIP; }

  // ── Réseau ──────────────────────────────────────────────────────────────────

  _openSocket() {
    this._socket = dgram.createSocket({ type: "udp4", reuseAddr: true });

    this._socket.on("message", (msg, rinfo) => {
      if (rinfo.address === this.myIP) return;
      this._handleMessage(msg.toString("utf8"), rinfo.address);
    });

    this._socket.on("error", (e) => {
      if (this.running) console.warn("[election] Socket error:", e.message);
    });

    this._socket.bind(ELECT_PORT, "0.0.0.0", () => {
      try { this._socket.addMembership(MCAST_ADDR, this.myIP); } catch (e) {
        try { this._socket.addMembership(MCAST_ADDR); } catch (e2) {}
      }
      console.log(`[election] Écoute ${MCAST_ADDR}:${ELECT_PORT}`);
    });
  }

  _broadcast(msg) {
    const buf = Buffer.from(msg, "utf8");
    const s   = dgram.createSocket("udp4");
    s.bind(0, () => {
      try { s.setMulticastInterface(this.myIP); } catch (e) {}
      s.setMulticastTTL(4);
      s.send(buf, 0, buf.length, ELECT_PORT, MCAST_ADDR, () => s.close());
    });
    console.log(`[election] → ${msg}`);
  }

  // ── Élection ────────────────────────────────────────────────────────────────

  _startElection() {
    if (!this.running) return;
    if (this._electionPending) return; // debounce : une seule élection à la fois
    this._electionPending = true;
    this.state = "CANDIDATE";
    console.log(`[election] Lancement (nodeId=${this.myNodeId})`);
    this._broadcastElection();

    clearTimeout(this._electionTimer);
    this._electionTimer = setTimeout(() => {
      this._electionPending = false;
      if (this.state === "CANDIDATE") this._becomeLeader();
    }, ELECTION_WAIT);
  }

  _broadcastElection() {
    const now = Date.now();
    if (now - this._lastBroadcastTs < BROADCAST_COOLDOWN) return; // anti-storm
    this._lastBroadcastTs = now;
    this._broadcast(`ELECTION:${this.myNodeId}:${this.myIP}`);
  }

  _becomeLeader() {
    if (!this.running) return;
    this.state    = "LEADER";
    this.leaderIP = this.myIP;
    console.log(`[election] LEADER élu: ${this.myIP}`);
    this._broadcast(`LEADER:${this.myNodeId}:${this.myIP}`);
    this.onBecomeLeader(this.myIP);
    this._stopWatchdog();
    this._startHeartbeat();
  }

  _becomeFollower(newLeaderIP) {
    if (!this.running) return;
    const changed = newLeaderIP !== this.leaderIP;
    // Annule toute élection en cours
    clearTimeout(this._electionTimer);
    this._electionPending = false;
    this.state    = "FOLLOWER";
    this.leaderIP = newLeaderIP;
    this.lastHeartbeat = Date.now();
    console.log(`[election] FOLLOWER — leader: ${newLeaderIP}`);
    if (changed) this.onLeaderElected(newLeaderIP);
    this._stopHeartbeat();
    this._startWatchdog();
  }

  // ── Traitement des messages ──────────────────────────────────────────────────

  _handleMessage(msg, senderIP) {
    const parts = msg.split(":");
    if (parts.length < 3) return;
    const [type, nodeIdStr, senderNode] = parts;
    const senderId = parseInt(nodeIdStr, 10);

    switch (type) {
      case "ELECTION":
        if (senderId > this.myNodeId) {
          // ID supérieur reçu → on se retire quelle que soit notre état (CANDIDATE ou LEADER)
          if (this.state === "LEADER") {
            // Un nœud supérieur arrive ou revient : on cède immédiatement
            console.log(`[election] ELECTION d'un supérieur (${senderId}) alors qu'on est LEADER — démission`);
            this._stopHeartbeat();
            this.leaderIP = null;
          }
          if (this.state === "CANDIDATE") {
            clearTimeout(this._electionTimer);
            this._electionPending = false;
          }
          this.state = "FOLLOWER";
          // Reset lastHeartbeat pour laisser le supérieur le temps de se proclamer
          this.lastHeartbeat = Date.now();
          this._startWatchdog(); // surveille que le supérieur se proclame bien
          // Protocole Bully : répondre OK pour signaler qu'on se déférence
          this._broadcast(`OK:${this.myNodeId}:${this.myIP}`);
        } else if (senderId < this.myNodeId) {
          // Notre ID est plus grand → on démarre notre propre élection (anti-storm)
          if (!this._electionPending) {
            this._startElection();
          } else {
            this._broadcastElection(); // déjà candidat, re-broadcast avec cooldown
          }
        }
        // senderId === myNodeId : collision d'IP improbable, ignoré
        break;

      case "OK":
        // Un nœud supérieur prend le relais — on annule notre candidature
        if (this.state === "CANDIDATE" && senderId > this.myNodeId) {
          clearTimeout(this._electionTimer);
          this._electionPending = false;
          this.state = "FOLLOWER";
          this.lastHeartbeat = Date.now();
          this._startWatchdog();
        }
        break;

      case "LEADER":
        console.log(`[election] LEADER reçu: ${senderNode} (nodeId=${senderId})`);
        if (senderId >= this.myNodeId) {
          // Le leader a un ID >= au nôtre — on se soumet (>= évite split-brain si IDs égaux)
          this._becomeFollower(senderNode);
        } else if (!this._electionPending) {
          // Notre ID est plus grand et pas d'élection en cours — on challenge
          console.log(`[election] LEADER inférieur (${senderId} < ${this.myNodeId}) — challenge`);
          this._startElection();
        }
        break;

      case "HEARTBEAT":
        if (senderNode === this.leaderIP) {
          // Heartbeat du leader connu : reset watchdog
          this.lastHeartbeat = Date.now();
        } else if (senderId > this.myNodeId) {
          // Heartbeat d'un nœud supérieur inconnu comme leader → on le reconnaît
          this._becomeFollower(senderNode);
        }
        // Heartbeat d'un nœud inférieur alors qu'on est LEADER → on ignore
        break;
    }
  }

  // ── Heartbeat ────────────────────────────────────────────────────────────────

  _startHeartbeat() {
    this._stopHeartbeat();
    this._heartbeatTimer = setInterval(() => {
      if (this.state === "LEADER" && this.running) {
        this._broadcast(`HEARTBEAT:${this.myNodeId}:${this.myIP}`);
      }
    }, HEARTBEAT_INTERVAL);
  }

  _stopHeartbeat() {
    clearInterval(this._heartbeatTimer);
    this._heartbeatTimer = null;
  }

  // ── Watchdog ─────────────────────────────────────────────────────────────────

  _startWatchdog() {
    this._stopWatchdog();
    this.lastHeartbeat = Date.now();
    this._watchdogTimer = setInterval(() => {
      if (this.state !== "FOLLOWER" || !this.running) return;
      const elapsed = Date.now() - this.lastHeartbeat;
      if (elapsed > LEADER_TIMEOUT) {
        console.warn(`[election] Leader silencieux depuis ${elapsed}ms → élection`);
        this._startElection();
      }
    }, LEADER_TIMEOUT / 2);
  }

  _stopWatchdog() {
    clearInterval(this._watchdogTimer);
    this._watchdogTimer = null;
  }
}

module.exports = { LeaderElection, getLocalIP };

File: dewicom-mobile/app/src/main/java/com/dewicom/LeaderElection.java
package com.dewicom;

import android.content.Context;
import android.net.wifi.WifiManager;
import android.util.Log;

import java.net.DatagramPacket;
import java.net.DatagramSocket;
import java.net.InetAddress;
import java.net.MulticastSocket;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.ScheduledFuture;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicLong;
import java.util.concurrent.atomic.AtomicReference;

/**
 * Élection de leader de type Bully sur multicast UDP.
 *
 * Protocole :
 *  - Chaque nœud a un nodeId = hash de son IP (long)
 *  - Le nœud avec le plus grand nodeId devient leader
 *  - Messages multicast sur ELECT_PORT :
 *      ELECTION:<nodeId>:<ip>  → je candidate
 *      LEADER:<nodeId>:<ip>    → j'ai gagné
 *      HEARTBEAT:<nodeId>:<ip> → je suis vivant (du leader, toutes les 2s)
 *
 * Si le leader ne heartbeate plus pendant LEADER_TIMEOUT_MS → nouvelle élection.
 */
public class LeaderElection {
    private static final String TAG = "LeaderElection";

    public static final String MCAST_ADDR = "224.0.0.251";
    public static final int ELECT_PORT   = 9998;

    private static final int HEARTBEAT_INTERVAL_MS = 2000;
    private static final int LEADER_TIMEOUT_MS      = 6000;
    private static final int ELECTION_WAIT_MS       = 2000; // attente de réponses ELECTION

    public enum State { FOLLOWER, CANDIDATE, LEADER }

    public interface Listener {
        /** Appelé quand ce nœud devient leader (doit démarrer le serveur). */
        void onBecomeLeader(String myIP);
        /** Appelé quand un leader est élu (pas nous) — doit se connecter à leaderIP. */
        void onLeaderElected(String leaderIP);
    }

    private final Context context;
    private final String myIP;
    private final long   myNodeId;
    private final Listener listener;

    private volatile State  state         = State.FOLLOWER;
    private volatile String leaderIP      = null;
    private final AtomicLong lastHeartbeat = new AtomicLong(0);
    private final AtomicReference<String> currentLeaderIP = new AtomicReference<>(null);
    private volatile boolean electionPending = false; // debounce : une seule élection à la fois

    private ScheduledExecutorService scheduler;
    private ScheduledFuture<?> heartbeatTask;
    private ScheduledFuture<?> watchdogTask;
    private ScheduledFuture<?> electionTask;
    private Thread listenThread;
    private volatile boolean running = false;

    private DatagramSocket sendSocket;
    private MulticastSocket recvSocket;
    private WifiManager.MulticastLock mcastLock;

    public LeaderElection(Context context, String myIP, Listener listener) {
        this.context  = context;
        this.myIP     = myIP;
        this.myNodeId = ipToNodeId(myIP);
        this.listener = listener;
        Log.d(TAG, "Init — IP: " + myIP + " nodeId: " + myNodeId);
    }

    // ── Lifecycle ─────────────────────────────────────────────────────────────

    public void start() {
        running   = true;
        scheduler = Executors.newScheduledThreadPool(2);

        acquireMulticastLock();
        startListening();

        // Lance une première élection après un délai aléatoire (évite les collisions)
        long delay = 500 + (long)(Math.random() * 1000);
        scheduler.schedule(this::startElection, delay, TimeUnit.MILLISECONDS);
    }

    public void stop() {
        running = false;
        electionPending = false;
        if (electionTask  != null) { electionTask.cancel(false);  electionTask  = null; }
        if (heartbeatTask != null) { heartbeatTask.cancel(false); heartbeatTask = null; }
        if (watchdogTask  != null) { watchdogTask.cancel(false);  watchdogTask  = null; }
        if (scheduler     != null) scheduler.shutdownNow();
        if (listenThread  != null) listenThread.interrupt();
        try { if (recvSocket != null) recvSocket.close(); } catch (Exception ignored) {}
        try { if (sendSocket != null) sendSocket.close(); } catch (Exception ignored) {}
        if (mcastLock != null && mcastLock.isHeld()) mcastLock.release();
    }

    public State  getState()    { return state; }
    public String getLeaderIP() { return currentLeaderIP.get(); }
    public boolean isLeader()   { return state == State.LEADER; }

    // ── Élection ──────────────────────────────────────────────────────────────

    private synchronized void startElection() {
        if (!running) return;
        if (electionPending) return; // debounce : une seule élection à la fois
        electionPending = true;
        state = State.CANDIDATE;
        Log.d(TAG, "Lancement élection (nodeId=" + myNodeId + ")");
        broadcast("ELECTION:" + myNodeId + ":" + myIP);

        if (electionTask != null) electionTask.cancel(false);
        electionTask = scheduler.schedule(() -> {
            electionPending = false;
            if (state == State.CANDIDATE) becomeLeader();
        }, ELECTION_WAIT_MS, TimeUnit.MILLISECONDS);
    }

    private synchronized void becomeLeader() {
        if (!running) return;
        state = State.LEADER;
        currentLeaderIP.set(myIP);
        Log.d(TAG, "Je suis le LEADER (" + myIP + ")");
        broadcast("LEADER:" + myNodeId + ":" + myIP);
        listener.onBecomeLeader(myIP);
        stopWatchdog();
        startHeartbeat();
    }

    private synchronized void becomeFollower(String newLeaderIP) {
        if (!running) return;
        boolean changed = !newLeaderIP.equals(currentLeaderIP.get());
        // Annule toute élection en cours
        if (electionTask != null) { electionTask.cancel(false); electionTask = null; }
        electionPending = false;
        state = State.FOLLOWER;
        currentLeaderIP.set(newLeaderIP);
        lastHeartbeat.set(System.currentTimeMillis());
        Log.d(TAG, "Je suis FOLLOWER, leader=" + newLeaderIP);
        if (changed) listener.onLeaderElected(newLeaderIP);
        stopHeartbeat();
        startWatchdog();
    }

    // ── Traitement des messages ────────────────────────────────────────────────

    private synchronized void handleMessage(String msg, String senderIP) {
        if (!running) return;
        String[] parts = msg.split(":");
        if (parts.length < 3) return;
        String type       = parts[0];
        long   senderId   = Long.parseLong(parts[1]);
        String senderNode = parts[2];

        switch (type) {
            case "ELECTION":
                if (senderId > myNodeId) {
                    // ID supérieur → annule notre candidature, reset heartbeat pour laisser
                    // le temps au supérieur de se proclamer avant que le watchdog intervienne
                    Log.d(TAG, "ELECTION d'un plus grand nodeId (" + senderId + ") — déférence");
                    if (state == State.CANDIDATE) {
                        if (electionTask != null) { electionTask.cancel(false); electionTask = null; }
                        electionPending = false;
                        state = State.FOLLOWER;
                    }
                    lastHeartbeat.set(System.currentTimeMillis());
                } else if (senderId < myNodeId && !electionPending) {
                    // Notre ID est plus grand → on s'annonce si pas déjà candidat
                    broadcast("ELECTION:" + myNodeId + ":" + myIP);
                }
                break;

            case "LEADER":
                Log.d(TAG, "LEADER reçu: " + senderNode + " (nodeId=" + senderId + ")");
                if (senderId >= myNodeId) {
                    // ID >= au nôtre → on se soumet (>= évite split-brain si IDs égaux)
                    becomeFollower(senderNode);
                } else if (!electionPending) {
                    // Notre ID est plus grand et pas d'élection en cours → on challenge
                    Log.d(TAG, "LEADER inférieur (" + senderId + " < " + myNodeId + ") — challenge");
                    startElection();
                }
                break;

            case "HEARTBEAT":
                if (senderNode.equals(currentLeaderIP.get())) {
                    // Heartbeat du leader connu → reset watchdog
                    lastHeartbeat.set(System.currentTimeMillis());
                } else if (senderId > myNodeId) {
                    // Heartbeat d'un nœud supérieur inconnu comme leader → on le reconnaît
                    becomeFollower(senderNode);
                }
                // Heartbeat d'un nœud inférieur (on est LEADER) → ignoré
                break;
        }
    }

    // ── Heartbeat (leader) ────────────────────────────────────────────────────

    private void startHeartbeat() {
        stopHeartbeat();
        heartbeatTask = scheduler.scheduleWithFixedDelay(
            () -> {
                if (state == State.LEADER) {
                    broadcast("HEARTBEAT:" + myNodeId + ":" + myIP);
                }
            },
            0, HEARTBEAT_INTERVAL_MS, TimeUnit.MILLISECONDS);
    }

    private void stopHeartbeat() {
        if (heartbeatTask != null) { heartbeatTask.cancel(false); heartbeatTask = null; }
    }

    // ── Watchdog (follower surveille le leader) ───────────────────────────────

    private void startWatchdog() {
        stopWatchdog();
        lastHeartbeat.set(System.currentTimeMillis());
        watchdogTask = scheduler.scheduleWithFixedDelay(() -> {
            if (state != State.FOLLOWER || !running) return;
            long elapsed = System.currentTimeMillis() - lastHeartbeat.get();
            if (elapsed > LEADER_TIMEOUT_MS) {
                Log.w(TAG, "Leader silencieux depuis " + elapsed + "ms → élection");
                startElection();
            }
        }, LEADER_TIMEOUT_MS, LEADER_TIMEOUT_MS / 2, TimeUnit.MILLISECONDS);
    }

    private void stopWatchdog() {
        if (watchdogTask != null) { watchdogTask.cancel(false); watchdogTask = null; }
    }

    // ── Réseau ────────────────────────────────────────────────────────────────

    private void broadcast(String msg) {
        try {
            if (sendSocket == null || sendSocket.isClosed()) {
                sendSocket = new DatagramSocket();
                sendSocket.setBroadcast(true);
            }
            byte[] data   = msg.getBytes("UTF-8");
            InetAddress group = InetAddress.getByName(MCAST_ADDR);
            DatagramPacket pkt = new DatagramPacket(data, data.length, group, ELECT_PORT);
            sendSocket.send(pkt);
            Log.d(TAG, "→ " + msg);
        } catch (Exception e) {
            Log.w(TAG, "Erreur broadcast: " + e.getMessage());
        }
    }

    private void startListening() {
        listenThread = new Thread(() -> {
            try {
                recvSocket = new MulticastSocket(ELECT_PORT);
                recvSocket.setReuseAddress(true);
                recvSocket.setSoTimeout(500);
                InetAddress group = InetAddress.getByName(MCAST_ADDR);
                recvSocket.joinGroup(group);
                Log.d(TAG, "Écoute élection sur " + MCAST_ADDR + ":" + ELECT_PORT);

                byte[] buf = new byte[256];
                while (running) {
                    try {
                        DatagramPacket pkt = new DatagramPacket(buf, buf.length);
                        recvSocket.receive(pkt);
                        String msg       = new String(pkt.getData(), 0, pkt.getLength(), "UTF-8");
                        String senderIP  = pkt.getAddress().getHostAddress();
                        if (!senderIP.equals(myIP)) { // ignore nos propres messages
                            handleMessage(msg, senderIP);
                        }
                    } catch (java.net.SocketTimeoutException ignored) {
                    } catch (Exception e) {
                        if (running) Log.w(TAG, "Erreur réception: " + e.getMessage());
                    }
                }
                recvSocket.leaveGroup(group);
                recvSocket.close();
            } catch (Exception e) {
                Log.e(TAG, "Erreur socket élection: " + e.getMessage());
            }
        }, "dewicom-election");
        listenThread.setDaemon(true);
        listenThread.start();
    }

    private void acquireMulticastLock() {
        try {
            WifiManager wm = (WifiManager) context.getApplicationContext()
                    .getSystemService(Context.WIFI_SERVICE);
            mcastLock = wm.createMulticastLock("dewicom_election");
            mcastLock.setReferenceCounted(true);
            mcastLock.acquire();
        } catch (Exception e) {
            Log.w(TAG, "MulticastLock: " + e.getMessage());
        }
    }

    // ── Utilitaires ───────────────────────────────────────────────────────────

    private static long ipToNodeId(String ip) {
        try {
            String[] p = ip.split("\\.");
            return ((long)(Integer.parseInt(p[0]) & 0xFF) << 24)
                 | ((long)(Integer.parseInt(p[1]) & 0xFF) << 16)
                 | ((long)(Integer.parseInt(p[2]) & 0xFF) <<  8)
                 | ((long)(Integer.parseInt(p[3]) & 0xFF));
        } catch (Exception e) {
            return (long)(Math.random() * Long.MAX_VALUE);
        }
    }
}
